### ISSUE TITLE
fetch.on('message') firing for messages outside fetched range

### ISSUE THREAD

**Issue by NeoPhi** [mscdex/node-imap#464]
Using a fetch command such as:

```
const fetch = connection.fetch('14201:14300', {bodies: ''});
fetch.on('message', (message, seqno) => {
  // handle message events
});
```

I'm seeing seqno come in that are not part of the requested range, if the connected to Box is undergoing change. It looks like the change events are being fired on the fetch as message events instead of the connection's update event. The debug output which includes my own output is:

```
<= '* 14005 FETCH (X-GM-THRID 900154593605437028 X-GM-MSGID 900154593605437028 X-GM-LABELS (Awesome) UID 14300 MODSEQ (1) INTERNALDATE "15-Mar-1997 19:31:57 +0000" FLAGS (\\Seen) )'
fetch:14201:14300:message:14005 ATTRIBUTES
fetch:14201:14300:message:14005 END
<= '* 350602 FETCH (UID 402782 MODSEQ (10000934) FLAGS (\\Seen))'
fetch:14201:14300 MESSAGE 350602 { domain: null, _events: {}, _maxListeners: undefined }
<= '* 350604 FETCH (UID 402784 MODSEQ (10000977) FLAGS (\\Seen))'
fetch:14201:14300 MESSAGE 350604 { domain: null, _events: {}, _maxListeners: undefined }
<= 'A6 OK Success'
=> 'IDLE IDLE'
fetch:14201:14300 END
<= '+ idling'
```

**Comment by mscdex** [comment]
Use the `seqno` namespace if you are using sequence numbers:

```javascript
const fetch = connection.seqno.fetch('14201:14300', {bodies: ''});
```

**Comment by NeoPhi** [comment]
I am requesting via UIDs.

**Comment by mscdex** [comment]
Can you also show the debug output for the actual fetch request too? Did you try specifying individual messages instead of using a range? Typically using UID ranges is not that useful because they are not necessarily incremented in order like sequence numbers, so it's entirely possible for an IMAP server to have a newer message with UID 100 and an older message with UID 101.

**Comment by NeoPhi** [comment]
This was the debug output for the start of the fetch command:

```
fetch:14201:14300 START
<= '+ idling'
=> DONE
<= 'IDLE OK IDLE terminated (Success)'
=> 'A6 UID FETCH 14201:14300 (X-GM-THRID X-GM-MSGID X-GM-LABELS MODSEQ UID FLAGS
 INTERNALDATE BODY.PEEK[])'
<= '* 13906 FETCH (X-GM-THRID 896734438572134286 X-GM-MSGID 896810014921880036 X
-GM-LABELS (Migrated) UID 14201 MODSEQ (1) INTERNALDATE "06-Feb-1997 21:31:18 +0
000" FLAGS (\\Seen) BODY[] {1270}'
fetch:14201:14300 MESSAGE 13906
fetch:14201:14300:message:13906 BODY
<= '* 13906 FETCH (X-GM-THRID 896734438572134286 X-GM-MSGID 896810014921880036 X
-GM-LABELS (Migrated) UID 14201 MODSEQ (1) INTERNALDATE "06-Feb-1997 21:31:18 +0
000" FLAGS (\\Seen) )'
```

I have not tried requesting individual UIDs. It was my understanding that provided the UIDVALIDITY of the mailbox didn't change UIDs are assigned in an ascending fashion as per RFC 3501.

**Comment by mscdex** [comment]
Ok, so after some more digging it [seems](http://stackoverflow.com/questions/26262472/gmail-imap-is-sometimes-returning-bad-results-for-fetch) Gimap quietly changed things late last year and are now sending flag updates pretty much anywhere. Ugh...

**Comment by mscdex** [comment]
Thinking about this some more, I'm not sure of the best way to work around this. At first I thought maybe we could just check if the message seqno/UID is something we requested (by checking against any ranges, exact seqnos/UIDs), but if a wildcard is used things get tricky because something like: "`3291:*` includes the UID of the last message in the mailbox, even if that value is less than 3291". So we would not know ahead of time whether that message is something we requested or not.

**Comment by NeoPhi** [comment]
It looks like these unsolicited updates can be detected by the fact that the modseq of the incoming message is higher than the highestmodseq returned by the server during the initial status if I'm understanding RFC 4551 correctly.

**Comment by mscdex** [comment]
That might work, but I do not know if it is safe to assume unsolicited fetches will always have a modseq value attached or if we can assume that the updated flags and the modseq will be contained within one single untagged fetch (I have seen Yahoo's IMAP implementation split up attribute pairs for a single message across multiple untagged (solicited) fetch responses).

**Comment by NeoPhi** [comment]
If I'm reading [RFC 7162](http://www.ietf.org/rfc/rfc7162.txt) correctly, I don't know if this completely answers the question, but it contains this clarification:

> Once a client issues a CONDSTORE enabling command, it has announced itself as a "CONDSTORE-aware client".  The server MUST then include mod-sequence data in all subsequent untagged FETCH responses (until the connection is closed), whether they were caused by a regular STORE, a STORE with an UNCHANGEDSINCE modifier, or an external agent.

**Comment by mscdex** [comment]
Right, that makes sense, but if the server doesn't support CONDSTORE can it still send unsolicited, untagged fetches at any time (just minus modseq attributes)?

**Comment by NeoPhi** [comment]
Looks like [RFC 3501](http://www.ietf.org/rfc/rfc3501.txt) section 7, says that this is possible:

> Server implementations that offer multiple simultaneous access to the same mailbox SHOULD also send appropriate unilateral untagged FETCH and EXPUNGE responses if another agent changes the state of any message flags or expunges any messages.

As a consumer of the current API I would be okay getting those messages. I think what tripped me up the most is that the returned message never fired the body, attributed, or end events.
